packages/server/src/router.ts

```diff
@@ -62,10 +62,12 @@ function getDataTransformer(
 
 export type inferHandlerInput<TProcedure extends Procedure> =
   TProcedure extends ProcedureWithInput<any, infer TInput, any>
-    ? undefined extends TInput
-      ? [(TInput | null | undefined)?]
-      : [TInput]
-    : [(undefined | null)?];
+    ? undefined extends TInput // ? is input optional
+      ? unknown extends TInput // ? is input unset
+        ? [(null | undefined)?] // -> there is no input
+        : [(TInput | null | undefined)?] // -> there is optional input
+      : [TInput] // -> input is required
+    : [(undefined | null)?]; // -> there is no input
```

Comment:
@danielyogel this is where the magic happens

Reply:
@KATT Looks good to me!

From what I understand, this solution is when no input required by server, the client still expects an optional `null | undefined` input, I would change that to `[]; // -> there is no input`, but I guess this is a breaking change, and also a very very minor issue...